server/src/main/java/org/elasticsearch/cluster/coordination/ElectionStrategy.java

```diff
@@ -105,10 +110,14 @@ public void beforeCommit(long term, long version, ActionListener<Void> listener)
         listener.onResponse(null);
     }
 
-    public boolean nodeMayWinElection(ClusterState lastAcceptedState, DiscoveryNode node) {
+    public NodeEligibility nodeMayWinElection(ClusterState lastAcceptedState, DiscoveryNode node) {
         final String nodeId = node.getId();
-        return lastAcceptedState.getLastCommittedConfiguration().getNodeIds().contains(nodeId)
+        if (lastAcceptedState.getLastCommittedConfiguration().getNodeIds().contains(nodeId)
             || lastAcceptedState.getLastAcceptedConfiguration().getNodeIds().contains(nodeId)
-            || lastAcceptedState.getVotingConfigExclusions().stream().noneMatch(vce -> vce.getNodeId().equals(nodeId));
+            || lastAcceptedState.getVotingConfigExclusions().stream().noneMatch(vce -> vce.getNodeId().equals(nodeId))) {
+            return new NodeEligibility(true, "");
```

Comment:
Since we now return the reason, it would be nice to split above condition and return different explanation for voting exclusion and `getLastCommittedConfiguration`. 

Also should we take into account shutdown entries as well? I realize it is tricky if somebody create them for all nodes at once. Probably better not to, for now.

Reply:
I think if we wanted to add a shutdown check -- crossed my mind, too -- we should open a new ticket for it. I can file a ticket, if you think it's worth doing.